File: easy.server.demo/easy.server.boot.demo/src/main/java/io/github/jinlongliao/easy/server/boot/demo/logic/service/TestAsyncService.java
package io.github.jinlongliao.easy.server.boot.demo.logic.service;

import io.github.jinlongliao.easy.server.cached.annotation.EnableCache;
import io.github.jinlongliao.easy.server.cached.annotation.WeAsync;
import io.github.jinlongliao.easy.server.utils.common.UUIDHelper;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Service;

import java.lang.invoke.MethodHandles;

/**
 * 测试异步
 *
 * @author: liaojinlong
 * @date: 2023/9/18 21:29
 */
@Service
@EnableCache
public class TestAsyncService {
    private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());

    @Async
    public void testAsync() {
        log.info("threadId:{} name:{}", Thread.currentThread().getId(), Thread.currentThread().getName());
    }

    @WeAsync(maxTimeout = 1000L)
    public Object testWeAsync(boolean block) {
        if (block) {
            try {
                Thread.sleep((20000));
            } catch (InterruptedException e) {
                log.error(e.getMessage(), e);
                return e.getMessage();
            }
        }
        log.info("threadId:{} name:{}", Thread.currentThread().getId(), Thread.currentThread().getName());
        return UUIDHelper.mongoId();
    }
}

File: easy.server.demo/easy.server.kotlin.demo/src/main/kotlin/io/github/jinlongliao/easy/server/kotlin/demo/logic/service/TestAsyncService.kt
package io.github.jinlongliao.easy.server.kotlin.demo.logic.service

import io.github.jinlongliao.easy.server.cached.annotation.EnableCache
import io.github.jinlongliao.easy.server.cached.annotation.WeAsync
import io.github.jinlongliao.easy.server.utils.common.UUIDHelper
import org.slf4j.LoggerFactory
import org.springframework.scheduling.annotation.Async
import org.springframework.stereotype.Service
import java.lang.invoke.MethodHandles

/**
 * 测试异步
 *
 * @author: liaojinlong
 * @date: 2023/9/18 21:29
 */
@Service
@EnableCache
class TestAsyncService {
    @Async
    fun testAsync() {
        log.info("threadId:{} name:{}", Thread.currentThread().id, Thread.currentThread().name)
    }

    @WeAsync(maxTimeout = 1000L)
    fun testWeAsync(block: Boolean): Any? {
        if (block) {
            try {
                Thread.sleep(20000.toLong())
            } catch (e: InterruptedException) {
                log.error(e.message, e)
                return e.message
            }
        }
        log.info("threadId:{} name:{}", Thread.currentThread().id, Thread.currentThread().name)
        return UUIDHelper.mongoId()
    }

    companion object {
        private val log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass())
    }
}
